### ISSUE TITLE
Adding to equals() hashCode() toString() 

### ISSUE THREAD

**Issue by danielshaya** [square/javapoet#378]
Something that would be super useful would be the ability to have default implementations for equals() hashCode() and toString().  In the same way the IDEs generated them for you.  It could be added to the TypeSpec object with methods such as generateToString().

**Comment by jbgi** [comment]
Probably more appropriate for a project like https://github.com/sormuras/poetry

**Comment by sormuras** [comment]
True. ;-)

**Comment by danielshaya** [comment]
I wasn't aware of this other project...
To be honest I think that having all the functions in one library makes it so much intuitive to use. I've used JavaPoet for a big project see my blog here (http://www.rationaljava.com/2015/12/7-tips-for-successful-code-generation.html) and apart from static imports it was perfect.  This suggested sort of added functionality would really take the project to the next level.

**Comment by jbgi** [comment]
I am currently writing an annotation processor (https://github.com/derive4j/derive4j) based on a plugable architecture and javapoet is part of the API of the contract between the main annotation processor and its plugins. As such I would rather prefer javapoet to not be bloated with frequently updated utilities which would require me to update my processor-api contract too frequently (so that plugins can take advantage of added functionalities).

On the contrary, a separated artifact like poetry is better: it can be used by plugins through shadowing without impacting the contract or other plugins (that may use a different/incompatible version of poetry).

But it is only my pov.

**Comment by danielshaya** [comment]
Sounds interesting but a corner case for JavaPoet.

I would have thought that most people who use JavaPoet are looking for a simple, clean, one-stop library to help them generate code. So generating default implementations equals() of hashcode() toString(), which is such a common task that IDEs support generation would be cool.  I have to say that having used JavaPoet on a fairly large project I can't think of anything else I would add at this stage.

But like you that only my pov.

**Comment by sormuras** [comment]
@danielshaya `import static` is in 1.5.0 - and on topic, i added your generator idea to: https://github.com/sormuras/javapoet/wiki - let's implement 'em!

**Comment by danielshaya** [comment]
Before I make a contribution can I ask with which version of the jdk are you compatible? Can I use Objects from Java 7?

**Comment by jbgi** [comment]
yes, javapoet targets jdk7+

**Comment by sormuras** [comment]
https://github.com/sormuras/poetry targets Java 8, btw.

**Comment by bjakke** [comment]
This would be super cool feature! +1

However as I'm currently using javapoet to generate code that must run on java 6 (works great), I would like to ask clarification on the "yes, javapoet targets jdk7+" comment above: I would assume this means that the comment would mean only that javapoet requires itself java 7+ to run (which it does), not that the generated code would require java 7. Also it would be nice if the minimum jre version for the generated code would be documented in the readme (at least I could not find it anywhere).

**Comment by jbgi** [comment]
@bjakke yes, there was a bit of a confusion. you are right that jdk7+ is needed by javapoet, not necessarily by the generated code, which I believe could be any java version that your compiler is able to target... and even more, using [backporting tools](https://en.wikipedia.org/wiki/Java_backporting_tools).

**Comment by bjakke** [comment]
Great!, in that case the above comment 

> Before I make a contribution can I ask with which version of the jdk are you compatible? Can I use Objects from Java 7?

would maybe need rethinking because I interpreted it as using the Objects class as for the implementation of the methods, i.e. generated code, which would make the generated code require java 7+ as Objects was added in 1.7 (it has lots of utils that would make making those methods easier). 

Still it would make a very nice feature for the most so I guess as long as method implementation generation is optional or can be turned off, it does not matter that much. Anyway thanks for making me aware of the backporting tools :)

**Comment by swankjesse** [comment]
FWIW, this is a good place to start:
https://github.com/square/wire/blob/d48be72904d7f6e1458b762cd936b1a7069c2813/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java#L1278-L1403
